client/buffer-pool.c

```diff
@@ -98,37 +91,33 @@ static void destroy_buffer(struct buffer *buffer) {
 	if (buffer->pango) {
 		g_object_unref(buffer->pango);
 	}
-	memset(buffer, 0, sizeof(struct buffer));
+	memset(buffer, 0, sizeof(struct pool_buffer));
 }
 
-struct buffer *get_next_buffer(struct window *window) {
-	struct buffer *buffer = NULL;
+struct pool_buffer *get_next_buffer(struct wl_shm *shm,
+		struct pool_buffer pool[static 2], uint32_t width, uint32_t height) {
+	struct pool_buffer *buffer = NULL;
 
-	int i;
-	for (i = 0; i < 2; ++i) {
-		if (window->buffers[i].busy) {
+	for (size_t i = 0; i < 2; ++i) {
+		if (pool[i].busy) {
 			continue;
 		}
-		buffer = &window->buffers[i];
+		buffer = &pool[i];
 	}
 
 	if (!buffer) {
 		return NULL;
 	}
 
-	if (buffer->width != window->width || buffer->height != window->height) {
+	if (buffer->width != width || buffer->height != height) {
 		destroy_buffer(buffer);
 	}
 
 	if (!buffer->buffer) {
-		if (!create_buffer(window, buffer,
-					window->width, window->height, window->scale,
+		if (!create_buffer(shm, buffer, width, height,
 					WL_SHM_FORMAT_ARGB8888)) {
 			return NULL;
 		}
 	}
-
-	window->cairo = buffer->cairo;
-	window->buffer = buffer;
 	return buffer;
 }
```

Comment:
I've removed the "window" abstraction but kept this file for wl_buffer management. Refactored the old "window" abstraction out of it.